File: Leetcode/tree/589_N-aryTreePreorderTraversal/Solution.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val,List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
    public List<Integer> preorder(Node root) {
        List<Integer> res = new ArrayList<>();
        dfs(root, res);
        return res;
    }
        private void dfs(Node root, List<Integer> res){
            if(root==null) return;
            res.add(root.val);
            if(root.children == null) return;
            for(Node n: root.children){
                dfs(n, res);
            }
        }
}

File: Leetcode/tree/589_N-aryTreePreorderTraversal/Solution.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, children):
        self.val = val
        self.children = children
"""
class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        res = []
        if root is None:return res
        res.append(root.val);
        for child in root.children:
            res.extend(self.preorder(child))
        return res
